File: subarray_sum_equals_k.cpp
/* Continous subarray = indices should be continous; 
   [1,2,3,4,5,6,7] k=9 ; [2,3,4] and [4,5] */
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int answer = 0;
        if(nums.size() == 0)
            return answer;
        unordered_map<int,int>hash_array;
        int sum = 0;
        for(int i=0;i<nums.size();i++)
        {
            sum += nums[i];
            if(sum == k)
                answer++;
            if(hash_array.count(sum-k))
                answer += hash_array[sum-k];
            
            hash_array[sum]++;
        }
        return answer;
    }
};

File: Fb_ph/Practice/Python/Subarray Sum Equals K.py
'''
Given an array of integers nums and an integer k, return the total number of subarrays whose sum equals to k.

Example 1:

Input: nums = [1,1,1], k = 2
Output: 2

Examnple 2:
Input: nums = [0,0,0,0,0,0,0,0] k= 0

'''
'''
TC:- O(n) where n is the length of the array
SC:- O(n) for hash Map
'''
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        frequency = {}
        Answer = 0;
        Sum = 0;
        for i in range(len(nums)):
            Sum += nums[i];
            if Sum == k :
                Answer += 1;
            if Sum-k in frequency:
                Answer += frequency[Sum-k];
            frequency[Sum] = frequency.get(Sum,0) + 1
        
        return Answer
    
